engine/src/utils/format.rs

```diff
@@ -16,23 +22,71 @@ where
         let width = self[0].as_ref().len() as u32;
         let height = self.len() as u32;
 
-        let data: Vec<u8> = self
-            .iter()
-            .flat_map(|slice| {
-                slice
-                    .as_ref()
+        let mut data = vec![];
+        {
```

Comment:
nit: extra block scope here is not needed. (`write_image`, takes `self` not `&mut self`).